src/Chainweb/Utils/Watchdog.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE CPP #-}
+
+module Chainweb.Utils.Watchdog (withWatchdog, notifyReady) where
+
+#ifdef WITH_SYSTEMD
+import Control.Concurrent
+import qualified Control.Concurrent.Async as Async
+import Control.Monad
+import System.Environment (lookupEnv)
+import qualified System.Systemd.Daemon as SD
+
+import Chainweb.Utils (tryAllSynchronous)
+
+withWatchdog :: IO a -> IO a
+withWatchdog act = do
+    usec <- maybe defaultWatchdogUsec read <$> lookupEnv "WATCHDOG_USEC"
+    Async.race act (watchdogTimer $! usec `div` 2) >>=
+        either return impossible
+  where
+    impossible _ = fail "impossible: watchdog returned first"
+
+    watchdogTimer :: Int -> IO ()
+    watchdogTimer usec0 = forever $ do
+        void $ tryAllSynchronous $ SD.notifyWatchdog
+        threadDelay usec
+      where
+        usec = if usec0 <= 0 then defaultWatchdogUsec else usec0
```

Comment:
Any potential issues here the `usec` is *just* too close to what the external `systemd` is expecting?

Reply:
In general no -- typical values for watchdog in seconds are 30s - 2m